data_analysis/pii/utils/evaluation.py

```diff
@@ -0,0 +1,103 @@
+import json
+
+TAGS = ['EMAIL', 'IP_ADDRESS', 'KEY']
+
+
+def load_json(sample):
+    try:
+        return json.loads(sample)
+    except ValueError:
+        return []
+
+
+def overlapped(a, b, alpha=0.8, beta=0.8):
+    """Returns True if the intervals a and b overlap for more than 80% of their lengths"""
+    size_overlap = max(0, min(a[1], b[1]) - max(a[0], b[0]))
+    ref_overlap = size_overlap / (b[1] - b[0])
+    pred_overlap = size_overlap / (a[1] - a[0])
+    return (ref_overlap > alpha and pred_overlap > beta)
+
+
+def compare_intervals(references, predictions, alpha=0.8, beta=0.8):
+    """Compare two lists of intervals and return the number 
+    of true positives, false positives and false negatives.
+    >>> compare_intervals([(0, 7), (10, 20)], [(1,8), (99, 119)], 0, 0)[0]
+    {'TP': 1, 'FN': 1, 'FP': 1}
+    """
+    ref_intervals = sorted(references, key=lambda x: x[0])
+    pred_intervals = sorted(predictions, key=lambda x: x[0])
+    scores = {"TP": 0, "FN": 0, "FP": 0}
+    detected_secrets = []
+    for interval in pred_intervals:
+        for target in ref_intervals:
+            if overlapped(interval, target, alpha, beta):
+                # the prediction is a true positive
+                scores["TP"] += 1
```

Comment:
TP_ref and TP_pred are not the same in some cases, for example:
- ref = [(0, 1), (2, 3)]
- pred = [(1, 10)]

Here TP_ref = 2, TP_pred = 1
For example, if a SSH key is very long and your regex only detect parts of it, it would be better to separate the two. However, it is not very comman

Reply:
I see, but I don't really get the intuition behind the separation because if we want to catch these cases we'd have to lower the overlap threshold to at least 0.5, so I think the results won't change with our current thresholds